### ISSUE TITLE
An overloaded function cannot call identical overloaded function with its parameters

### ISSUE THREAD

**Issue by mckravchyk** [microsoft/TypeScript#53318]
# Bug Report

### 🔎 Search Terms

no overload matches call  implementation succeed

### 🕗 Version & Regression Information

At least in 4.9 and it does occur in the current nightly

This is the behavior in every version I tried, and I reviewed the FAQ for entries.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230317#code/GYVwdgxgLglg9mABDMYCmAnA4mqAKASgC5EAlNCODAEwB4BnKDFAcwBpFHmwWA+AbgBQoSLATJUmHPhjUSXVsU5NWiAD6IwIADbahI6PCQp02XHln15KngG0AukvKUaDG+2XcW6zTu0DhcENxEylzWQB5DABJanoAfmsvHwU7RxJnKjpUjxyfLV1eRABvQUEAXzKDMSQ4ECgw-CcKLLcvDhyA6qNEOoazGTlPRSTVDQK9QNEevsaLONG05pds9w73fL8uoJre+rnImLjE4Z4U9wdl1pz15PGtksFERAxcEAxjSQH5qNj6An4iAA9EDEAA5OC9ABumG0cAAhtREABbeFQCAACzQ9EQUAxMBxEHhugAdBUgA)

### 💻 Code

```ts
function innerGet(): Record<string, string>;
function innerGet(id: string): string | null;
function innerGet(ids: string[]): Record<string, string | null>;
function innerGet(idOrIds?: string | string[]): Record<string, string | null> {

}

function outerGet(): Record<string, string>;
function outerGet(id: string): string | null;
function outerGet(ids: string[]): Record<string, string | null>;
function outerGet(idOrIds?: string | string[]): Record<string, string | null> {
  return innerGet(idOrIds);
  // No overload matches this call.
  // The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.
}
```

### 🙁 Actual behavior

Trying to call a function with the same overloaded  parameters will result in a "no overload matches this call" error.

### 🙂 Expected behavior

There should be no error. It's pretty clear that the types are compatible.

**Comment by MartinJohns** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/39885 / https://github.com/microsoft/TypeScript/issues/1805.

**Comment by mckravchyk** [comment]
Ok, thanks.  I never came across this in 3 years of using TS and I have been using plenty overloads. It's good to know that, better late then never but I wished the docs mentioned it. This is an issue that will not come up immediately, so there may be plenty of code using overloads that will need to be refactored upon realization that there is a major bug that has been marked as wontfix. I only found: "Always prefer parameters with union types instead of overloads when possible"  but I don't think it sufficiently warns against it. I would interpret it as "use union types if you don't need to narrow the return type based on the input arg".

And when I try to implement the same with the extends keyword, I find myself between the rock and a hard place. I run into a bug I encountered earlier #52051 . As a matter of fact, my workaround to that bug was using overloads :)


```ts
function get<Arg extends string | string[] | undefined>(
  idOrIds: Arg
) : Arg extends string ? string | null
  : Arg extends string[] ? Record<string, string | null>
  : Record<string, string> {
  if (typeof idOrIds === 'string') {
    return ''; // Not assignable
  }
    
}

function get2<Arg extends string | string[] | undefined>(
  idOrIds: Arg
) : Arg extends string ? string | null
  : Arg extends string[] ? Record<string, string | null>
  : Record<string, string> {
  if (typeof idOrIds === 'string') {
    return '' as ReturnType<typeof get2>; // Not going to pass either
  }
    
}
```

[Playground](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230317#code/GYVwdgxgLglg9mABAcwKZQDwEEBOzGoAeUqYAJgM6IVQ4xj4A+1t9yA2gLqLPhmrB6qMgD4AFAChEiGGQDyOAJKUAXIlzIJASkRqNBYqUos6DRAH4TbHojAgANvam71eAyXJUapjt0sAlVAg4HDIMbzYAGiszZjtHEWc1QODQ8NYGaIiGEUQAb2cYYEQxKABPAAdUOGLZBWUqAF5mxABybORWnQLpaRx0EBwkVtaAbkQAegnEADk4KEQAQwoKGGQwRYAje1RnAF9naQkDiVBIWAQUdAAmbDciD2MOmw6uGz4BIVFJaTqlVVcmh0enuhk8MXwlmecQcTmkIPwDyMXgyvgsiBSITCHSyqJs8XsiXhGKCWPSPlxPlyPRkxVKlWqtXk-yaLXaqK6+UOiH6UEGw1aSyogT5QwAKgyMOUqjUrlBriJxlNZvMUHBrFA4IgKssqKgYFAABaoHD7bnHIA)

I guess I will use @ts-ignore, unless there's some way to override it with casting.
